pkg/ccl/changefeedccl/parquet_sink_cloudstorage.go

```diff
@@ -0,0 +1,254 @@
+	pqcs := &parquetCloudStorageSink{}
+
+	pqcs.wrapped = baseCloudStorageSink
+
+	return pqcs
+}
+
+// EmitRow does not do anything. It must not be called. It is present so that
+// parquetCloudStorageSink implements the Sink interface.
+func (pqcs *parquetCloudStorageSink) EmitRow(
+	ctx context.Context,
+	topic TopicDescriptor,
+	key, value []byte,
+	updated, mvcc hlc.Timestamp,
+	alloc kvevent.Alloc,
+) error {
+	return errors.Errorf("Emit Row should not be called for parquet format")
+}
+
+// Close implements the Sink interface.
+func (pqcs *parquetCloudStorageSink) Close() error {
+	return pqcs.wrapped.Close()
+}
+
+// Dial implements the Sink interface.
+func (pqcs *parquetCloudStorageSink) Dial() error {
+	return pqcs.wrapped.Dial()
+}
+
+// EmitResolvedTimestamp does not do anything as of now. It is there to
+// implement Sink interface.
+func (pqcs *parquetCloudStorageSink) EmitResolvedTimestamp(
+	ctx context.Context, encoder Encoder, resolved hlc.Timestamp,
+) error {
+	return errors.Errorf("parquet format does not support emitting resolved timestamp")
+}
+
+// Flush implements the Sink interface.
+func (pqcs *parquetCloudStorageSink) Flush(ctx context.Context) error {
+	return pqcs.wrapped.Flush(ctx)
+}
+
+// EncodeAndEmitRow links the buffer in the cloud storage sync file and the
+// parquet writer (see parquetCloudStorageSink). It also takes care of encoding
+// and emitting row event to cloud storage. Implements the SinkWithEncoder
+// interface.
+func (pqcs *parquetCloudStorageSink) EncodeAndEmitRow(
+	ctx context.Context,
+	updatedRow cdcevent.Row,
+	topic TopicDescriptor,
+	updated, mvcc hlc.Timestamp,
+	alloc kvevent.Alloc,
+) error {
+	s := pqcs.wrapped
+	var pqww *parquetWriterWrapper
+	file, err := s.getOrCreateFile(topic, mvcc)
+	file.alloc.Merge(&alloc)
+	if err != nil {
+		return err
+	}
+
+	if file.pw == nil {
+		var err error
+		pqww, err = makeparquetWriterWrapper(ctx, updatedRow, &file.buf)
+		if err != nil {
+			return err
+		}
+		file.pw = pqww
+	} else {
+		pqww = file.pw
+	}
+
+	i := -1
+	parquetRow := make(map[string]interface{}, pqww.numCols)
+	if err := updatedRow.ForEachColumn().Datum(func(d tree.Datum, col cdcevent.ResultColumn) error {
+		i++
+		// Omit NULL columns from parquet row (cannot think of any other way of
+		// encoding NULL values)
+		if d == tree.DNull {
+			parquetRow[col.Name] = nil
+			return nil
+		}
+		encodeFn, err := pqww.parquetColumns[i].GetEncoder()
+		if err != nil {
+			return err
+		}
+		edNative, err := encodeFn(d)
+		if err != nil {
+			return err
+		}
+
+		parquetRow[col.Name] = edNative
+
+		return nil
+
+	}); err != nil {
+		return err
+	}
+
+	if err = pqww.parquetWriter.AddData(parquetRow); err != nil {
+		return err
+	}
+
+	if pqww.parquetWriter.CurrentRowGroupSize() > s.targetMaxFileSize {
+		s.metrics.recordSizeBasedFlush()
+
+		if err = pqww.parquetWriter.Close(); err != nil {
```

Comment:
we'll need to close the writer if we return early at any point above, right? Put this in a defer?